ramips: add support for Hi-Link HLK-7688A IoT Gateway Mode

ewpa opened the pull request:
Please can support for Hi-Link HLK-7688A in IoT Gateway Mode be added.

adschm commented on target/linux/ramips/mt76x8/base-files/etc/board.d/01_leds:
```diff
@@ -42,6 +42,9 @@ hilink,hlk-7628n|\
 	set_wifi_led "$boardname:green:wlan"
 	;;
+hilink,hlk-7688a-gateway)
+	set_wifi_led "hlk-7688a:green:wlan"
```
Please use DT trigger instead.

adschm commented on target/linux/ramips/dts/mt7628an_hilink_hlk-7688a-gateway.dts:
```diff
@@ -0,0 +1,108 @@
+		bootargs = "console=ttyS0,57600";
+	};
+
+	buttonkeys {
```
keys

adschm commented on target/linux/ramips/dts/mt7628an_hilink_hlk-7688a-gateway.dts:
```diff
@@ -0,0 +1,108 @@
+	leds {
+		compatible = "gpio-leds";
+
+		wlan: status {
```
So, which?

adschm commented on target/linux/ramips/dts/mt7628an_hilink_hlk-7688a-gateway.dts:
```diff
@@ -0,0 +1,108 @@
+&spi0 {
+	status = "okay";
+
+	m25p80@0 {
```
flash@0

adschm commented on target/linux/ramips/dts/mt7628an_hilink_hlk-7688a-gateway.dts:
```diff
@@ -0,0 +1,108 @@
+	m25p80@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <10000000>;
```
Can this go faster?

ewpa replied:
U-Boot runs the SPI clock at 32 MHz.  Practical limit seems to be around 66 MHz.

adschm commented on target/linux/ramips/dts/mt7628an_hilink_hlk-7688a-gateway.dts:
```diff
@@ -0,0 +1,108 @@
+				label = "ALL";
+				reg = <0x0 0x02000000>;
+				read-only;
+			};
```
Please remove that, unless there is a specific reason for having it.

ewpa replied:
Factory firmware had ALL partition.  Easily remove.

adschm commented on target/linux/ramips/dts/mt7628an_hilink_hlk-7688a-gateway.dts:
```diff
@@ -0,0 +1,108 @@
+				read-only;
+			};
+
+			partition@1 {
```
partition@0

adschm commented on target/linux/ramips/dts/mt7628an_hilink_hlk-7688a-gateway.dts:
```diff
@@ -0,0 +1,108 @@
+
+/ {
+	compatible = "hilink,hlk-7688a-gateway", "hilink,hlk-7688a", "mediatek,mt7628an-soc";
+	model = "Hi-Link HLK-7688A IoT Gateway";
```
I'm not a big fan of creating one image per "mode". I'd prefer to create one image hilink,hlk-7688a or hilink,hlk-7688a-iot for both modes.

ewpa replied:
Okay, let me think about this some more.

ewpa replied:
Do you have any tips on the best method to disable the switch when the board is in "IoT Device Mode", i.e. with a single Ethernet (instead of the 1x WAN and 4x LAN of the "IoT Gateway Mode")?

adschm replied:
While I don't really understand what's that supposed to achieve (why would I have a separate mode which just provides less functionality, I think this is something to be achieved with config. OpenWrt does not aim at providing the exact mirror of the vendor concept _by default_, but just some reasonable configuration. So, you can just flash the official image twice and configure each of them based on the configuration you want to get.

ewpa replied:
By disabling some of the Ethernet ports, the pins can be used for other purposes, such as I2C or SD Card.

adschm commented on target/linux/ramips/dts/mt7628an_hilink_hlk-7688a-gateway.dts:
```diff
@@ -0,0 +1,108 @@
+
+	chosen {
+		bootargs = "console=ttyS0,57600";
+	};
```
This node can be dropped.

adschm commented on target/linux/ramips/dts/mt7628an_hilink_hlk-7688a-gateway.dts:
```diff
@@ -0,0 +1,108 @@
+		compatible = "gpio-keys";
+
+		wdt_rst {
+			label = "wdt_rst";
```
Reset?

981213 commented on target/linux/ramips/dts/mt7628an_hilink_hlk-7688a-gateway.dts:
```diff
@@ -0,0 +1,108 @@
+	compatible = "hilink,hlk-7688a-gateway", "hilink,hlk-7688a", "mediatek,mt7628an-soc";
+	model = "Hi-Link HLK-7688A IoT Gateway";
+
+	memory@0 {
```
memory node can be dropped. linux has the ability to detect it correctly.

ewpa commented:
I have resolved all the above issues in the amended commit 3ae29b6.  Do let me know if that was not the correct method to update this pull request.  Note the HLK-7688A IoT Device mode does boot with this configuration, but is not fully tested.

adschm commented on target/linux/ramips/image/mt76x8.mk:
```diff
@@ -118,6 +118,15 @@ define Device/hilink_hlk-7628n
 TARGET_DEVICES += hilink_hlk-7628n
 
+define Device/hilink_hlk-7688a
+  BLOCKSIZE := 64k
```
Why is that needed?

ewpa replied:
I will remove.

adschm commented on target/linux/ramips/dts/mt7628an_hilink_hlk-7688a.dts:
```diff
@@ -0,0 +1,93 @@
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <66000000>;
```
Speed > 50 MHz typically requires `m25p,fast-read;`

ewpa replied:
Thanks, I missed that.

adschm commented:
> Signed-off-by: Ewan Parker <34030942+ewpa@users.noreply.github.com>

Is that address used by intention?

Apart from that, you didn't move the LED trigger to DTS so far.
I'm also missing ethernet/esw setup in DTS, including the MAC address.
Please also add an overview of MAC addresses to the commit message, more details here: https://openwrt.org/docs/guide-developer/device-support-policies#mac_addresses

ewpa commented:
I plan to procure another board to check consistency of MAC address positions in the flash.

ewpa commented:
I have amended the commit again and hope to have fixed everything that was outstanding.  In this commit I also enabled (and tested) I2C and SPI CS1.  When testing with another board I needed to lower the SPI speed to 40 MHz.

adschm commented on target/linux/ramips/dts/mt7628an_hilink_hlk-7688a.dts:
```diff
@@ -0,0 +1,118 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/dts-v1/;
```
drop dts-v1, this has been moved to mt7628an.dtsi

adschm commented on target/linux/ramips/dts/mt7628an_hilink_hlk-7688a.dts:
```diff
@@ -0,0 +1,118 @@
+		compatible = "gpio-leds";
+
+		led_wlan: wlan {
+			label = "hlk-7688a:green:wlan";
```
please drop the model: prefix here, just use "green:wlan"

adschm commented on target/linux/ramips/dts/mt7628an_hilink_hlk-7688a.dts:
```diff
@@ -0,0 +1,118 @@
+		compatible = "winbond,w25q256", "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <40000000>;
+		m25p,fast-read;
```
I don't think m25p,fast-read is required below 50 MHz?

ewpa replied:
Agreed, I can see no benefit in retaining that under 50 MHz.

adschm commented on target/linux/ramips/dts/mt7628an_hilink_hlk-7688a.dts:
```diff
@@ -0,0 +1,118 @@
+};
+
+&wmac {
+	mtd-mac-address = <&factory 0x4>;
```
If you use the default location for caldata, this address is read automatically.

adschm commented on target/linux/ramips/mt76x8/base-files/etc/board.d/02_network:
```diff
@@ -218,6 +219,7 @@ ramips_setup_macs()
+	hilink,hlk-7688a|\
 	wiznet,wizfi630s)
 		label_mac=$(mtd_get_mac_binary factory 0x4)
 		wan_mac=$(mtd_get_mac_binary factory 0x28)
```
Setting wan to 0x28 does not make any sense, as you already have set &ethernet to that address in DTS ...

ewpa replied:
My mistake, WAN MAC is at offset 0x2e.  On my devices the factory MACs stored at both 0x28 and 0x2e happen to be the same value.

ewpa commented:
I have pushed commit 2009e3162c to address all concerns.

adschm commented on target/linux/ramips/dts/mt7628an_hilink_hlk-7688a.dts:
```diff
@@ -0,0 +1,115 @@
+		led-boot = &led_wlan;
+		led-failsafe = &led_wlan;
+		led-upgrade = &led_wlan;
+	};
```
Please move aliases block before keys block.

ewpa replied:
Will do.

adschm commented on target/linux/ramips/dts/mt7628an_hilink_hlk-7688a.dts:
```diff
@@ -0,0 +1,115 @@
+
+&state_default {
+	spis {
+		groups = "spis";
```
node name should be gpio instead of spis, and groups should be wdt and wlan_an IMO?

ewpa replied:
I will change to `wdt` and `wled_an`.

adschm commented on target/linux/ramips/dts/mt7628an_hilink_hlk-7688a.dts:
```diff
@@ -0,0 +1,115 @@
+	pinctrl-0 = <&spi_pins>, <&spi_cs1_pins>;
+
+	flash@0 {
+		compatible = "winbond,w25q256", "jedec,spi-nor";
```
Is the first entry required here?

ewpa replied:
Not strictly required, happy to get rid of the winbond reference.

ewpa committed:
ramips: add support for Hi-Link HLK-7688A

Specifications:

  - SoC: MediaTek MT7688AN
  - RAM: 128 MB
  - Flash: 32 MB
  - Ethernet: 5x 10/100 (1x WAN, 4x LAN)
  - Wireless: built in 2.4GHz (bgn)
  - USB: 1x USB 2.0 port
  - Buttons: 1x Reset
  - LEDs: 1x (WiFi)

Flash instructions:

  - Configure TFTP server with IP address 10.10.10.3
  - Name the firmware file as firmware.bin
  - Connect any Ethernet port to the TFTP server's LAN
  - Choose option 2 in U-Boot
  - Alternatively choose option 7 to upload firmware to the built-in
    web server

MAC addresses as verified by OEM firmware:

  use   address   source
  2g    *:XX      factory 0x4
  LAN   *:XX+1    factory 0x28
  WAN   *:XX+1    factory 0x2e

Notes:

This board is ostensibly a module containing the MediaTek MT7688AN SoC,
128 MB DDR2 SDRAM and 32 MB flash storage.  The SoC can be operated in
IoT Gateway Mode or IoT Device Mode.

From some vendors the U-Boot that comes installed operates on UART 2
which is inaccessible in gateway mode and operates unreliably in the
Linux kernel when using more than 64 MB of RAM.  For those, updating
U-Boot is recommended.

Signed-off-by: Ewan Parker <ewan@ewan.cc>

ewpa commented:
I have pushed amended commit e7b237d6b2 to address the last three issues.

alex9x commented:
Hi! Does someone know, why don`t work software reset on HLK-7688A?
4.xx have patch for 32mb flash with 0x66 0x99 sequence to back flash in 3byte address mode.
18.06.9 works well, but master or 19.x.x based on 5.x kernel fail to make soft reset.
Any help are welcome!

ewpa commented:
> Hi! Does someone know, why don`t work software reset on HLK-7688A?
> 18.06.9 works well, but master or 19.x.x based on 5.x kernel fail to make soft reset.

I also experienced this on earlier commits on master.  Using this commit and kernel and reboot works (for me currently):

_OpenWrt SNAPSHOT, r14641-e7b237d6b2
Linux 5.4.69 #0 Sat Oct 31 09:29:54 2020 mips GNU/Linux_

adschm commented:
Thanks, will be merged later today.
